### ISSUE TITLE
Add search/replace functionality?

### ISSUE THREAD

**Issue by 3noch** [ndmitchell/shake#266]
String search/replace is not part of Shake or Haskell's standard libraries. Thus, in order to do this simple task, you need to pull in a whole new dependency. I'm using `Text` to do it, but Text doesn't play nicely with Shake's functions since they are all `String`-based. Would it be possible to add a helper for this? A pure `String -> String -> String -> String` would be sufficient.

**Comment by ndmitchell** [comment]
I recommend using the extra library which provides Data.List.Extra.replace. That's already a dependency of Shake, so has no impact on your transitive dependencies. Generally, I'm trying to avoid adding things that aren't at least slightly build-system specific.

**Comment by 3noch** [comment]
Ok, sounds good.

**Comment by 3noch** [comment]
I've been confused as to whether your goal is primarily to provide a build-system *library* or a build system *tool*. While the two are very close, I can see them diverging in regards to "batteries-included" or not. The *tool* side would prefer as little extra-dependencies as possible (especially if the build system is not building Haskell code), but the *library* side prefers to promote composition as much as possible. I wonder if some clarity in this regard would be helpful for issue-tracking.

**Comment by 3noch** [comment]
I have just recently transitioned from using Shake more as a tool into more of a library. I'm now using `stack`/`cabal` to track dependencies and shake is the engine. I added a `.cabal` file, a `stack.yaml` file, and a small `.bat` script to wrap up shake's power behind a minimal-dependency interface. Since I've already brought `stack` into my build system, I can pull in any dependency I want, but the principles still hold for either side.

**Comment by ndmitchell** [comment]
I guess I'm slicing on a different divide than library vs tool. I'm happy to provide extra functionality, even if it's more general in nature, if it's disproportionately used by build systems, the existing Haskell stuff doesn't fit well or if there are advantages from integrating something a bit more closely. As an example, with `cmd` I can integrate with profiling. For things like `replace`, there's no real benefits to having it in the library, other than removing a dependency, and it's not that common in build systems (for example, I've used templating libraries more than I've used `replace`, perhaps as an alternative to that). But in truth, I'm still working it out myself.
